irpc.py

```diff
@@ -35,17 +36,31 @@ def hoist_declaration(main: FuncDef,
                     main.insert(0, node)
                     break
 
+def hoist_prov_memo(main: FuncDef,
+                    provdef: ProvDef):
+
+    _name_ = f"{e}_provided"
```

Comment:
You don't import pass variable `e` in this function.
You get just lucky with the strange python scope.

Reply:
I combined this function with your other hoisting function. Figured it would be easier and less costly than passing a lot of the same information to two separate functions